apps/builder/app/canvas/collapsed.ts

```diff
@@ -246,14 +231,25 @@ const recalculate = () => {
   // If most elements are collapsed at the next step, scrollHeight becomes equal to clientHeight,
   // which resets the scroll position. To prevent this, we set the document's height to the current scrollHeight
   // to preserve the scroll position.
-  const preserveHeight = document.documentElement.style.height;
-
-  // Mark that we are in the process of recalculating collapsed elements
-  document.documentElement.style.setProperty(
-    markCollapsedMutationProperty,
-    `true`
-  );
-  document.documentElement.style.height = `${document.documentElement.scrollHeight}px`;
+  // use nested element to avoid full page repaint and freeze on big projects
+  let collapsedElement = document.querySelector(
+    "#ws-collapsed"
```

Comment:
could probably just memoize the reference somewhere instead of dom query

Reply:
HMR will loose it most of the time